packages/vm/src/imports.rs

```diff
@@ -29,6 +33,16 @@ const MAX_LENGTH_CANONICAL_ADDRESS: usize = 32;
 /// The maximum allowed size for bech32 (https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32)
 const MAX_LENGTH_HUMAN_ADDRESS: usize = 90;
 const MAX_LENGTH_QUERY_CHAIN_REQUEST: usize = 64 * KI;
+
+/// Length of a (sha-256) hash in bytes
+const LENGTH_SHA256_HASH: usize = 32;
+
+/// Max length of a serialized signature
+const MAX_LENGTH_SIGNATURE: usize = 64;
+
+/// Max length of a serialized public key
+const MAX_LENGTH_PUBKEY: usize = 33;
```

Comment:
We should get more explicit about what pubkey types we accept. If we want to accept both compressed/uncompressed we should check:

1. Is uncompressed? (length = 65 and starts with 0x04)
2. Is compressed? (length = 33 and starts with 0x02 or 0x03)
3. Otherwise fail.

3. is important to avoid accidentally supporting other types, that people start relying on.

Reply:
I missed the uncompressed case here, thanks.

This will fail nevertheless, at  the low level (deserialization). But given that there are two different formats / lengths, it's a good idea to do a check, and fail early.

Reply:
There seem to be other prefixes than 0x02/0x03/0x04, which we are not aware of right now:

> or even the identity 0x00

https://github.com/RustCrypto/elliptic-curves/issues/291#issuecomment-771760187

Reply:
Yes... I'm thinking, instead of adding all this logic early, wouldn't be better to just read the maximum possible size (65 bytes), and delegate error handling / supported formats to the low level impl? It's more robust in the end.

Reply:
> or even the identity 0x00

That would be uncompressed, right?

What happens to read region if it reads out of bounds? We can always pad on writing, by the way.

Reply:
Right, increase the limit to 65 bytes and move the check into

```rust
pub fn secp256k1_verify(
    message_hash: &[u8],
    signature: &[u8],
    public_key: &[u8],
) -> VmResult<bool> {
```

Otherwise we had to check `public_key: &[u8]` again there.

Reply:
> That would be uncompressed, right?

I don't know how 0x00 works. I just get the impression we should only explicitly allow what we understand.

> What happens to read region if it reads out of bounds? We can always pad on writing, by the way.

The contract is free to send us broken Regions. We either get garbage as long as the memory region is part of the linear memory. Or `read_region` errors with a deref error.

Reply:
In fact, why not letting `k256::VerifyingKey` deserializer directly do the checking? That way we'll automatically support new formats, corner cases, etc.

Reply:
> In fact, why not letting `k256::VerifyingKey` deserializer directly do the checking? That way we'll automatically support new formats, corner cases, etc.

Because at some point in the not too far future we want to be able to write a spec for this interface, that is more explicit than _whatever k256 does_. Also we want to be able to switch the implementation to one that only supports 0x02/0x03/0x04.

Reply:
I would make as little assumptions as possible (basically, maximum possible length) and let the low level impl do the heavy lifting, checking, etc. That way, we just ride the support they offer, and don't introduce potential incompatibilities.

Reply:
I'll add a couple of tests for unsupported formats / prefixes (wrong length, wrong prefix, wrong pubkey), so that we get the idea of their failure modes, and associated error messagees.